File: personal/joe/hard/braille-joe/bit2letter.py
def get_header():
	return """#pragma once
#ifndef byte
#define byte unsigned char
#endif

byte letter2code(char letter)
{
  switch (letter)
  {
"""

def format(linenumber, content):
	output = '    '

	if content != 'null':
		flag = False
		for it in map(lambda x: x.rstrip(), content.split(' ')):
			if len(it) <= 1:
				if it != '\'':
					output += 'case \'%s\': ' % (it)
				else:
					output += 'case \'\\\'\': '
				flag = True
		if flag:
			output += 'return %d; break;\n' % (linenumber)

	return output

def close_header():
	return """
    default: return '?';
  }
}"""

if __name__ == '__main__':
	b2l = 'bit2letter'
	inlet = open(b2l + '.txt', 'r')
	outlet = open(b2l + '.h', 'w')

	i = 0
	outlet.write(get_header())
	for i, line in enumerate(inlet):
		it = format(i, line)
		if len(it.rstrip()) > 0:
			outlet.write(it)
		i += 1
	outlet.write(close_header())

	inlet.close()
	outlet.close()

File: personal/joe/hard/braille-joe/bit2letter.rb
def get_header
    crunchbang = '#'
    <<-header
#{crunchbang}pragma once
#{crunchbang}ifndef byte
#{crunchbang}define byte unsigned char
#{crunchbang}endif

byte letter2code(char letter)
{
  switch (letter)
  {
    header
end

def close_header
    <<-header
    
    default: return '?';
  }
}
    header
end

def get_code letter
    flag = false
    code = ""

    if letter.length <= 1
        code = (letter != "\'")? "case \'#{letter}\': " : "case \'\\\'\': "
        flag = true
    end

    return code, flag
end

def format line, content
    output = "    "

    if content
        flag = false
        for bit in content.split " "
            bit, ok = get_code bit
            output += bit if ok
            flag = flag || ok
        end
        output += "return #{line}; break;" if flag
    end

    return output
end

def main inlet, outlet
    i = 0
    outlet.puts get_header
    for line in inlet
        it = format i, line
        outlet.puts it if it.rstrip.length > 0
        i += 1
    end
    outlet.puts close_header
end

b2l = "bit2letter"
inlet = File.open(b2l + ".txt", "r")
outlet = File.open(b2l + ".h", "w")
main inlet, outlet
inlet.close
outlet.close
